components/raftstore-v2/src/operation/command/admin/merge/commit.rs

```diff
@@ -0,0 +1,652 @@
+        admin.mut_commit_merge().set_source(self.region().clone());
+        admin.mut_commit_merge().set_commit(state.get_commit());
+        admin.mut_commit_merge().set_entries(entries.into());
+        request.set_admin_request(admin);
+        // Please note that, here assumes that the unit of network isolation is store
+        // rather than peer. So a quorum stores of source region should also be the
+        // quorum stores of target region. Otherwise we need to enable proposal
+        // forwarding.
+        let msg = PeerMsg::AskCommitMerge(request);
+        // If target peer is destroyed, life.rs is responsible for telling us to
+        // rollback.
+        match store_ctx.router.force_send(target_id, msg) {
+            Ok(_) => (),
+            Err(SendError(PeerMsg::AskCommitMerge(msg))) => {
+                store_ctx
+                    .router
+                    .force_send_control(StoreMsg::AskCommitMerge(msg))
+                    .unwrap_or_else(|e| {
+                        slog_panic!(
+                            self.logger,
+                            "fails to send `AskCommitMerge` msg to store";
+                            "error" => ?e,
+                        )
+                    });
+            }
+            _ => unreachable!(),
+        }
+        Ok(())
+    }
+
+    pub fn on_reject_commit_merge(&mut self, index: u64) {
+        warn!(self.logger, "target peer rejected commit merge"; "index" => index);
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn on_ask_commit_merge<T>(
+        &mut self,
+        store_ctx: &mut StoreContext<EK, ER, T>,
+        req: RaftCmdRequest,
+    ) {
+        match self.validate_commit_merge(&req) {
+            Some(true) => {
+                let (ch, _) = CmdResChannel::pair();
+                self.on_admin_command(store_ctx, req, ch);
+            }
+            Some(false) => {
+                let commit_merge = req.get_admin_request().get_commit_merge();
+                let source_id = commit_merge.get_source().get_id();
+                let _ = store_ctx.router.force_send(
+                    source_id,
+                    PeerMsg::RejectCommitMerge {
+                        index: commit_merge.get_commit(),
+                    },
+                );
+            }
+            _ => (),
+        }
+    }
+
+    fn validate_commit_merge(&self, req: &RaftCmdRequest) -> Option<bool> {
+        let expected_epoch = req.get_header().get_region_epoch();
+        let source_region = req.get_admin_request().get_commit_merge().get_source();
+        let region = self.region();
+        if util::is_epoch_stale(expected_epoch, region.get_region_epoch()) {
+            info!(
+                self.logger,
+                "reject commit merge because of stale";
+                "current_epoch" => ?region.get_region_epoch(),
+                "expected_epoch" => ?expected_epoch,
+            );
+            Some(false)
```

Comment:
Maybe we can check if the source peer id is in `merged_records` first. If it's in `merged_records`, then it means the source should be destroyed instead of trying to rollback.